feat: Add Deck.gl Contour Layer

Mattc1221 opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

Adds the Contour Layer from deck.gl [here](https://deck.gl/docs/api-reference/aggregation-layers/contour-layer)

Related SIP: https://github.com/apache/superset/discussions/23553

_**Design Decisions:**_

The addition of the Contour layer follows the structure of existing deck.gl layers to maintain readability and consistency within the codebase.

_Structure_:

The directory `superset-frontend/plugins/legacy-preset-chart-deckgl/src/layers/Contour/` contains the Front end structure for the contour layer visualization.

- `controlPanel` configures the form to grab relevant information specified by the user to create and tune the contour layer.
- `Contour.tsx` contains the functions that define the Contour Layer and get relevant data.
- `index.ts` serves as the wrapper for the contour component and provides a metadata function to grab information to describe what this visualization is/does.

The contour layer requires a unique input of "contours". Contours are defined as either `Isolines` or `Isobands` (more information [here](https://deck.gl/docs/api-reference/aggregation-layers/contour-layer)). 
- `Isolines` require three fields, `threshold`, `color`, `strokeWidth`. `threshold` defines a value to separate aggregated values on the Map. `strokeWidth` defines the width of the line in pixels that separates these values. `color` is just the color of that line.
- `Isobands` are similar to `Isolines`, except the `threshold` takes 2 values to create polygons on the map instead of lines.

To account for this unique structure, I added a new control panel input component: 
![Screenshot 2023-05-20 at 11 38 33 PM](https://github.com/apache/superset/assets/36670322/0876a86d-f7f8-4925-a8aa-4889745be575)
The added files for this input component are located under `superset-frontend/src/explore/components/controls/ContourControl/`:

- `types.ts` contains all the types needed for this new component.
- `index.tsx` controls the stateful variables that control adding/editing/removing contours.
- `ContourOption.tsx` displays added contours and an appropriate tooltip.
- `ContourPopoverControl.tsx` defines a popup with appropriate inputs for creating/editing an `isoline` and `isoband`.
- `ContourPopoverTrigger.tsx` wraps the `ContourPopoverControl` and controls when it is displayed.

The Back End code for this feature follows the back end code for other deck.gl visualizations as well, providing functions to structure queried data for the Heatmap component.


### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

New Visualization Recording:

https://github.com/apache/superset/assets/36670322/6b8da315-23e7-4ed9-a0ce-f0905815a405

<details>

<summary>Changing order of contours</summary>

https://github.com/apache/superset/assets/36670322/118b97f3-7334-4041-9392-5f5171a57996

</details>


### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

_**Set up**_
1. Start Superset
2. Click on the `Charts` tab at the top of the screen
3. Click the `+ Chart` button at the top right
4. Load a dataset with geo-positional data (longitude/latitude)
5. Search the charts for `Deck.gl Contour` or click on the `Map` chart category
6. Select `deck.gl Contour` and hit `Create New Chart`
7. Update the `Longitude & Latitude` field to the correct columns
8. Select a `Weight` function
9. Select `Create Chart`
10. There should now be a contour layer present on the screen
    - Note: the underlying map component may not be visible if the MAP_BOX_API_KEY is not set. To set it follow the instructions [here](https://superset.apache.org/docs/frequently-asked-questions/#why-is-the-map-not-visible-in-the-geospatial-visualization)
    - Note: It may look like the contour layer is not working if the `Cell Size` parameter is not large enough. For example, using the `flights` dataset, you should increase `Cell Size` to 50000.

Check that the following cause an immediate update
- Check that increasing/decreasing the `Cell Size` control 
- Try adding a new isoline by clicking `Click to add new contour`
- Try adding a new isoband by clicking `Click to add new contour`
- Try clicking an existing contour option and ensure that a popover appears with the appropriate fields pre-filled
- Delet an existing contour option

Other Checks:
- Check that hovering a contour option and check that the tooltip shows appropriate information
- Check that in the contour popover, the submit button is disabled when:
    - Not all fields are filled
    - All fields are filled, but a non-integer is entered in any of the textfields
- Check all descriptions make sense
- Check that changing the order of contours (by drag and drop) updates their z-index.
   - If there are overlapping contours, contours at the top of the list may be hidden by contours beneath them.
   - _Note: z-fighting may occur depending on how many contours are defined/how many are displayed. However this is an issue with the library itself._
 

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Mattc1221 committed:
feat: add contour layers

Mattc1221 committed:
fix: close contour menu on save/close

Mattc1221 committed:
fix: remove duplicate tooltip in contour list

Mattc1221 committed:
fix: updates to contour popover trigger

Mattc1221 committed:
fix: fix typo

Mattc1221 committed:
chore: package updates

Mattc1221 committed:
feat: open contour popover from contour option

Mattc1221 committed:
fix: autofill contour values with defaults when editing

Mattc1221 committed:
feat: update current contour on edit

Mattc1221 committed:
fix: updated thumbnails

Mattc1221 committed:
fix: persist changes and update position of new contour popover

Mattc1221 committed:
chore: clean up imports

Mattc1221 committed:
feat: validation in contour popover

Mattc1221 committed:
feat: replace cell size with slider and render trigger for cell size and contours

Mattc1221 committed:
fix: convert cell size to text field instead of slider

Mattc1221 committed:
chore: fix typescript linting and warnings

Mattc1221 committed:
fix: run prettier formatter

Mattc1221 committed:
Merge branch 'master' into deck_gl_contour_layer

Mattc1221 committed:
fix: deck.gl types

Mattc1221 commented:
@kgabryje, @rusackas and @villebro

Since you reviewed https://github.com/apache/superset/pull/23551 I was wondering if you would take a look at this PR as well?

Mattc1221 commented:
@villebro @rusackas @kgabryje @justinpark @michael-s-molina

I was wondering what an ETA on the timeline for review would be? Thanks again!

villebro commented:
> @villebro @rusackas @kgabryje @justinpark @michael-s-molina
> 
> I was wondering what an ETA on the timeline for review would be? Thanks again!

Hi @Mattc1221 , sorry for the delay and thanks for the ping. I'll put this on my review list for today.

Mattc1221 commented:
> > @villebro @rusackas @kgabryje @justinpark @michael-s-molina
> 
> > 
> 
> > I was wondering what an ETA on the timeline for review would be? Thanks again!
> 
> 
> 
> Hi @Mattc1221 , sorry for the delay and thanks for the ping. I'll put this on my review list for today.

Thank you @villebro, I really appreciate it 😃

villebro commented on docker/pythonpath_dev/superset_config.py:
```diff
@@ -108,6 +108,8 @@ class CeleryConfig(object):
 
 SQLLAB_CTAS_NO_LIMIT = True
 
+MAPBOX_API_KEY = os.getenv("MAPBOX_API_KEY")
+
```
We appear to be using `get_env_variable` in this file, maybe use that here, too, for consistency?
```suggestion
MAPBOX_API_KEY = get_env_variable("MAPBOX_API_KEY")

```

villebro commented on superset-frontend/src/explore/components/controls/ContourControl/ContourPopoverControl.tsx:
```diff
@@ -0,0 +1,311 @@
+  errorMapType,
+} from './types';
+
+enum CONTOUR_TYPES {
+  ISOLINE = 'ISOLINE',
+  ISOBAND = 'ISOBAND',
+}
```
nit: PascalCase tends to be used for enums, at least in the official TS docs: https://www.typescriptlang.org/docs/handbook/enums.html#string-enums

```suggestion
enum CONTOUR_TYPES {
  Isoline = 'ISOLINE',
  Isoband = 'ISOBAND',
}
```

villebro commented on superset-frontend/src/explore/components/controls/TextControl/index.tsx:
```diff
@@ -30,7 +30,7 @@ export interface TextControlProps<T extends InputValueType = InputValueType> {
   isFloat?: boolean;
   isInt?: boolean;
-  onChange?: (value: T, errors: any) => {};
+  onChange?: (value: T, errors: any) => void;
```
good catch, thanks for improving this 👍

villebro reviewed (commented):
A few minor comments. Let me put aside some time to see what could be done about migrating these to the v1 API, as it hurts my soul to see us adding LOCs to `viz.py`.. 😆

Mattc1221 committed:
fix: replace os.getevn with get_env_variable

Mattc1221 committed:
fix: replace uppercase enums with PascalCase

Mattc1221 committed:
fix: update remaing uppercase enum references

Mattc1221 committed:
Merge branch 'master' into deck_gl_contour_layer

Mattc1221 committed:
fix: type definition typo

Mattc1221 commented:
@villebro Thanks for the review! I believe this PR is ready for re-review :)

Mattc1221 commented:
Hey @villebro! I'm wondering if you would be able to take another pass through this PR?

Mattc1221 commented:
Hello @villebro! I was just wondering what the status update is for this PR? All of the checks are passing and the requested changes have been made!

JohnDietrich-Pepper commented:
@villebro @kgabryje @rusackas @michael-s-molina  any chance someone could take a look at this.  Would be nice to have this functionality and it appears that it has been ready for re-review for about a month and a half now.

kgabryje commented:
@Mattc1221 @JohnDietrich-Pepper Thank you for ping and sorry for long delay - I'll review the PR by tomorrow

kgabryje commented:
/testenv up

kgabryje commented on superset-frontend/plugins/legacy-preset-chart-deckgl/src/layers/Contour/controlPanel.ts:
```diff
@@ -0,0 +1,126 @@
+      validators: [validateNonEmpty],
+    },
+  },
+  formDataOverrides: formData => ({
```
Shouldn't that be 
```
formDataOverrides: formData => ({
    ...formData,
    size: getStandardizedControls().shiftMetric(),
  }),
```
?

It's used for retaining the control values when you switch the viz type, and it basically tells the control panel to prefill "size" control with the value of metric from the previous viz.

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/ContourOption.tsx:
```diff
@@ -0,0 +1,101 @@
+
+const ColorPatch = styled.div<{ formattedColor: string }>`
+  background-color: ${({ formattedColor }) => formattedColor};
+  height: 5px;
```
Could we use grid units from the theme here? Like `${({ theme }) => theme.gridUnit }px;`. 1 grid unit is 4px and we prefer using whole units, so it's going to be 4px instead of 5 but I guess that shouldn't make much difference here

Mattc1221 replied:
Thanks for the pointer! I wasn't aware of theme grid units!

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/ContourOption.tsx:
```diff
@@ -0,0 +1,101 @@
+    : 'undefined';
+
+  const formatIsoline = (threshold: number, width: any) =>
+    `Threshold: ${threshold}, color: ${formattedColor}, stroke width: ${width}`;
```
Could you wrap that in translation (`t(...)`)? Same 2 lines below

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/ContourOption.tsx:
```diff
@@ -0,0 +1,101 @@
+    ? `rgba(${color.r}, ${color.g}, ${color.b}, 1)`
+    : 'undefined';
+
+  const formatIsoline = (threshold: number, width: any) =>
```
Could you add type for width? I suppose it's either a number or string?

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/ContourOption.tsx:
```diff
@@ -0,0 +1,101 @@
+  const overlay = (
+    <div className="contour-tooltip-overlay">
+      <StyledListItem>
+        Threshold:{' '}
```
Translation here and lines 72, 77 and 80

Mattc1221 replied:
I've updated the overlay with translations (below). I'm not sure where to add translations on line 72. The result here is just numbers, do those need to be translated as well?

```.tsx
    <div className="contour-tooltip-overlay">
      <StyledListItem>
        {t("Threshold: ")}
        {isIsoband
          ? `[${lowerThreshold}, ${upperThreshold}]`
          : `${lowerThreshold}`}
      </StyledListItem>
      <StyledListItem>
        {t("Color: ")}<ColorPatch formattedColor={formattedColor} /> {t(`${formattedColor}`)}
      </StyledListItem>
      {!isIsoband && (
        <StyledListItem>{t(`Stroke Width: ${strokeWidth}`)}</StyledListItem>
      )}
    </div>
```

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/ContourOption.tsx:
```diff
@@ -0,0 +1,101 @@
+  return (
+    <ContourPopoverTrigger saveContour={saveContour} value={contour}>
+      <StyledOptionWrapper
+        key={index}
```
Is the key required here?

Mattc1221 replied:
Good catch! this is not required

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/ContourPopoverControl.tsx:
```diff
@@ -0,0 +1,311 @@
+
+const StyledRow = styled(Row)`
+  width: 100%;
+  gap: 10px;
```
Theme grid unit here 🙏

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/ContourPopoverControl.tsx:
```diff
@@ -0,0 +1,311 @@
+            hovered
+          />
+          <ColorPickerControl
+            value={typeof contour === 'object' && contour?.color}
```
In other components we're not checking for contour's type so I'm assuming it's not necessary here?

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/ContourPopoverControl.tsx:
```diff
@@ -0,0 +1,311 @@
+        <Col flex="1">
+          <ControlHeader
+            name="isoband-threshold-lower"
+            label="Lower Threshold"
```
Please add translations for labels and descriptions in isobandSection and isolineSection 🙏

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/ContourPopoverControl.tsx:
```diff
@@ -0,0 +1,311 @@
+        <Tabs.TabPane
+          className="adhoc-filter-edit-tab"
+          key={CONTOUR_TYPES.Isoline}
+          tab="Isoline"
```
Translations for tab names too

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/index.tsx:
```diff
@@ -0,0 +1,143 @@
+ */
+
+import React, { useState, useEffect } from 'react';
+import styled from '@emotion/styled';
```
PLease import styled from superset-ui/core

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/index.tsx:
```diff
@@ -0,0 +1,143 @@
+`;
+
+const ContourControl = ({ onChange, ...props }: ContourControlProps) => {
+  const [popoverVisible, setpopoverVisible] = useState(false);
```
```suggestion
  const [popoverVisible, setPopoverVisible] = useState(false);
```

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/index.tsx:
```diff
@@ -0,0 +1,143 @@
+      />
+    ));
+
+  const ghostButtonText = 'Click to add a contour';
```
translation

kgabryje commented on superset-frontend/src/explore/components/controls/ContourControl/types.ts:
```diff
@@ -0,0 +1,55 @@
+import { OptionValueType } from 'src/explore/components/controls/DndColumnSelectControl/types';
+import { ControlComponentProps } from 'src/explore/components/Control';
+
+export interface colorType {
```
Could you use PascalCase for types?

kgabryje commented on superset-frontend/src/explore/components/controls/DndColumnSelectControl/types.ts:
```diff
@@ -32,6 +32,7 @@ export interface OptionProps {
   isExtra?: boolean;
   datasourceWarningMessage?: string;
   canDelete?: boolean;
+  tooltipOverlay?: any;
```
```suggestion
  tooltipOverlay?: ReactNode;
```

kgabryje commented on superset-frontend/plugins/legacy-preset-chart-deckgl/src/layers/Contour/Contour.tsx:
```diff
@@ -0,0 +1,97 @@
+
+  const contours = rawContours?.map(
+    (contour: {
+      color: any;
```
Can we use ColorType here instead of any?

kgabryje commented:
I think changing the contour type between isoline and isoband doesn't work well.
1. I can change the contour type from isoline to isoband, but I can't do the opposite - when I edit an isoband, change the tab to isoline and type value, it only edits the lower threshold of isoband. I think user could expect that the contour becomes an isoline.
2. The popover should close after clicking save
3. Maybe we should have some kind of validation to ensure that upper threshold is larger than lower threshold?

https://github.com/apache/superset/assets/15073128/d00d4266-c512-435f-9b61-fea0aba5428f

kgabryje commented:
Oh 1 more thing that I noticed - there isn't any tooltip with values when I hover over contours in the chart. Maybe it'd be worth to add it?

kgabryje commented on superset-frontend/plugins/legacy-preset-chart-deckgl/src/layers/Contour/controlPanel.ts:
```diff
@@ -0,0 +1,126 @@
+              renderTrigger: true,
+              choices: [
+                ['sum', t('sum')],
+                ['mean', t('mean')],
```
deck.gl also allows min and max, should we add them too?

kgabryje commented on superset-frontend/plugins/legacy-preset-chart-deckgl/src/layers/Contour/Contour.tsx:
```diff
@@ -0,0 +1,97 @@
+    cellSize: Number(cellSize || '200'),
+    aggregation: aggregation.toUpperCase(),
+    getPosition: (d: { position: number[]; weight: number }) => d.position,
+    getWeight: (d: { weight: any }) => d.weight || 0,
```
```suggestion
    getWeight: (d: { weight: number }) => d.weight || 0,
```

Mattc1221 committed:
Merge branch 'master' into deck_gl_contour_layer

Mattc1221 committed:
fix: retain control values

Mattc1221 committed:
fix: add translations

Mattc1221 committed:
fix: PascalCase for Types and update case for setPopoverVisible

Mattc1221 committed:
fix: use theme grid units

Mattc1221 committed:
fix: update types

Mattc1221 committed:
fix: remove redundant type check

Mattc1221 committed:
fix: add min and max options

Mattc1221 committed:
fix: close popover on save

Mattc1221 committed:
fix: popover change type and validation

Mattc1221 commented on superset-frontend/src/explore/components/controls/ContourControl/index.tsx:
```diff
@@ -0,0 +1,143 @@
+
+  return (
+    <>
+      <DndSelectLabel
```
Sounds good! I'm going to resolve this comment for now then, if you think it should be handled in this PR feel free to unresolve it and ping me!

Mattc1221 committed:
fix: tooltip

Mattc1221 committed:
fix: stroke width validation

Mattc1221 committed:
fix: linting errors

Mattc1221 commented:
Thanks @kgabryje for the detailed review, I believe I've addressed all of your comments!

> * I can change the contour type from isoline to isoband, but I can't do the opposite - when I edit an isoband, change the tab to isoline and type value, it only edits the lower threshold of isoband. I think user could expect that the contour becomes an isoline.
> * The popover should close after clicking save
> * Maybe we should have some kind of validation to ensure that upper threshold is larger than lower threshold?

You should now be able to change isobands to isolines, the Save button closes the form, I've added some validation messages and disabled saving when the lower threshold is greater than or equal to the upper threshold. The tooltip should now appear with the lat/long coordinates as well as the threshold for whatever contour the mouse is hovering.

https://github.com/apache/superset/assets/36670322/02e7cb41-183b-4932-95c8-1a51bcddfa1a

Mattc1221 committed:
fix: python misc test

kgabryje commented:
Thanks for updates @Mattc1221! 1 last thing - for some reason, the inputs unfocus after typing 1 character, which makes typing more than 1 digit difficult. I didn't notice that happening before the latest changes

https://github.com/apache/superset/assets/15073128/e91b218e-db16-41ce-97c6-e5a64c0045b2

Mattc1221 committed:
Merge branch 'master' into deck_gl_contour_layer

Mattc1221 committed:
Merge branch 'master' into deck_gl_contour_layer

Mattc1221 committed:
fix: keep focus after typing

Mattc1221 commented:
@kgabryje I'm sorry about the delay!

The last comment on the unfocusing issues should now be fixed. I had pulled out the tabs sections into functions within the component. This was causing them to get re-rendered on every change and thus lose focus. I've put them inside of the return statement of the surrounding component

https://github.com/apache/superset/assets/36670322/9e463eb0-bd39-4cef-b947-42b3ce932f57

Mattc1221 commented:
@villebro @kgabryje @rusackas @justinpark @michael-s-molina I was wondering if one of you could take a final pass through this PR. All of the previous comments have been addressed and the PR should be ready to go.

kgabryje reviewed (approved):
LGTM! Thank you @Mattc1221 for the hard work... and patience 😄
